packages/kas/src/lexer.ts

```diff
@@ -0,0 +1,192 @@
+    ["\\!", {kind: "!"}], // not yet interpreted
+    ["<=|>=|<>|<|>|=", (value) => ({kind: "SIGN", value})],
+    ["\\\\le", {kind: "SIGN", value: "<="}],
+    ["\\\\ge", {kind: "SIGN", value: ">="}],
+    ["\\\\leq", {kind: "SIGN", value: "<="}],
+    ["\\\\geq", {kind: "SIGN", value: ">="}],
+    ["=\\/=", {kind: "SIGN", value: "<>"}],
+    ["\\\\ne", {kind: "SIGN", value: "<>"}],
+    ["\\\\neq", {kind: "SIGN", value: "<>"}],
+    ["\u2260", {kind: "SIGN", value: "<>"}], // ne
+    ["\u2264", {kind: "SIGN", value: "<="}], // le
+    ["\u2265", {kind: "SIGN", value: ">="}], // ge
+    ["\\\\frac", {kind: "FRAC"}],
+    ["\\\\dfrac", {kind: "FRAC"}],
+    ["sqrt|\\\\sqrt", {kind: "sqrt"}],
+    ["abs|\\\\abs", {kind: "abs"}],
+    ["ln|\\\\ln", {kind: "ln"}],
+    ["log|\\\\log", {kind: "log"}],
+    ["sin|cos|tan", (value) => ({kind: "TRIG", value})],
+    ["csc|sec|cot", (value) => ({kind: "TRIG", value})],
+    ["sinh|cosh|tanh", (value) => ({kind: "TRIG", value})],
+    ["csch|sech|coth", (value) => ({kind: "TRIG", value})],
+    ["\\\\sin", {kind: "TRIG", value: "sin"}],
+    ["\\\\cos", {kind: "TRIG", value: "cos"}],
+    ["\\\\tan", {kind: "TRIG", value: "tan"}],
+    ["\\\\csc", {kind: "TRIG", value: "csc"}],
+    ["\\\\sec", {kind: "TRIG", value: "sec"}],
+    ["\\\\cot", {kind: "TRIG", value: "cot"}],
+    ["\\\\arcsin", {kind: "TRIG", value: "arcsin"}], // should this be TRIGINV
+    ["\\\\arccos", {kind: "TRIG", value: "arccos"}], // should this be TRIGINV
+    ["\\\\arctan", {kind: "TRIG", value: "arctan"}], // should this be TRIGINV
+    ["\\\\arccsc", {kind: "TRIG", value: "arccsc"}], // should this be TRIGINV
+    ["\\\\arcsec", {kind: "TRIG", value: "arcsec"}], // should this be TRIGINV
+    ["\\\\arccot", {kind: "TRIG", value: "arccot"}], // should this be TRIGINV
+    ["arcsin|arccos|arctan", (value) => ({kind: "TRIGINV", value})],
+    ["arccsc|arcsec|arccot", (value) => ({kind: "TRIGINV", value})],
+    ["\\\\sinh", {kind: "TRIG", value: "sinh"}],
+    ["\\\\cosh", {kind: "TRIG", value: "cosh"}],
+    ["\\\\tanh", {kind: "TRIG", value: "tanh"}],
+    ["\\\\csch", {kind: "TRIG", value: "csch"}],
+    ["\\\\sech", {kind: "TRIG", value: "sech"}],
+    ["\\\\coth", {kind: "TRIG", value: "coth"}],
+    ["pi", {kind: "CONST", value: "pi"}],
+    ["\u03C0", {kind: "CONST", value: "pi"}], // pi
+    ["\\\\pi", {kind: "CONST", value: "pi"}],
+    ["theta", {kind: "VAR", value: "theta"}],
+    ["\u03B8", {kind: "VAR", value: "theta"}], // theta
+    ["\\\\theta", {kind: "VAR", value: "theta"}],
+    ["phi", {kind: "VAR", value: "phi"}],
+    ["\u03C6", {kind: "VAR", value: "phi"}], // phi
+    ["\\\\phi", {kind: "VAR", value: "phi"}],
+    ["[a-zA-Z]", (value) => ({kind: "VAR", value})], // TODO: post-process
+    [".", {kind: "INVALID"}],
+];
+
+type PreparedRule = [RegExp, Token | ((match: string) => Token)];
+
+const preparedRules: PreparedRule[] = rules.map((rule) => {
+    const regex = new RegExp("^(?:" + rule[0] + ")");
```

Comment:
So each rule becomes a regex that's anchored to the beginning of the input string and is non-capturing. Am I reading this correctly?

Reply:
Yes, that's correct.  It's a simplified version of https://github.com/zaach/jison-lex/blob/master/regexp-lexer.js#L10-L68 which is what `jison` uses internally for its lever.

Reply:
I don't think the non-capturing group part is really necessary since we only ever look at the whole string that's matched. 🤷‍♂️